tst/testinstall/MatrixObj/testmatobj_new.g

```diff
@@ -0,0 +1,299 @@
+                String( ex.nrCols ), ", ",
+                String( ex.mat ), " )", "\n\n"));
+            else 
+                Append(errorMsg, Concatenation(String(arg), "\n\n"));
+            fi;
+        od;
+        Add(errors, errorMsg);
+    fi;
+end;
+
+MatObjTest_CallFunc := function(f, args, breakOnError, errors, ex)
+    local catch;
+    #if f = ConstructingFilter then 
+    #    Error();
+    #fi;
+    if breakOnError then 
+        return CallFuncList(f, args);
+    else
+        catch := MatObjTest_CallAndCatchError(f, args);
+        if catch[1][1] then 
+            if IsBound(catch[1][2]) then 
+                return catch[1][2];
+            else 
+                return;
+            fi;
+        else 
+            MatObjTest_AppendErrorFail(f, args, catch, errors, ex);
+            return fail;
+        fi;
+    fi;
+end;
+
+MatObjTest_GenerateExample := function(filter, opt)
+    local inv, dims, dim, doms, dom, customGen, examples;
+    
+    examples := [];
+
+    if IsBound(opt.dimensions) then 
+        dims := opt.dimensions;
+    else 
+        dims := [[2,2],[3,3],[2,5],[5,5],[1,3],[4,1],[6,2]];
+    fi;
+
+    if IsBound(opt.domains) then 
+        doms := opt.domains;
+    else
+        doms := [Integers, Rationals, GF(2), GF(5), GF(49)];
+    fi;
+
+    if IsBound(opt.useCustomGenerator) then 
+        if IsBound(opt.invertible) then
+            customGen := opt.invertible;
+            for dim in dims do 
+                for dom in doms do 
+                    if dim[1] = dim[2] then 
+                        Add(examples, [customGen(filter, opt.randomSource, dim[1], dom), dim[1], dim[2], dom]);
+                    fi;
+                od;
+            od;
+        else 
+            customGen := opt.useCustomGenerator;
+            for dim in dims do 
+                for dom in doms do 
+                    Add(examples, [customGen(filter, opt.randomSource, dim[1], dim[2], dom), dim[1], dim[2], dom]);
+                od;
+            od;
+        fi; 
+    else 
+        if IsBound(opt.invertible) then 
+            for dim in dims do 
+                for dom in doms do 
+                    if dim[1] = dim[2] then 
+                        Add(examples, [RandomInvertibleMat(opt.randomSource, dim[1], dom), dim[1], dim[2], dom]);
+                    fi;
+                od;
+            od;
+        else 
+            for dim in dims do 
+                for dom in doms do 
+                    Add(examples, [RandomMat(opt.randomSource, dim[1], dim[2], dom), dim[1], dim[2], dom]);
+                od;
+            od;
+        fi;
+    fi;
+
+    if IsBound(opt.sourceOfTruth) then 
+        return List(examples, x-> 
+            rec(matObj := NewMatrix(filter, x[4], x[3], x[1]),
+            filter := filter,
+            baseDomain := x[4],
+            nrCols := x[3],
+            nrRows := x[2],
+            mat := x[1],
+            sourceOfTruth := NewMatrix(opt.sourceOfTruth, x[4], x[3], x[1]))
```

Comment:
As a reminder from our video chat: Note how `sourceOfTruth` has two different meanings here (once it is a matrix, once it is a filter)

Reply:
The filter specifying the sourceOfTruth is now called sourceOfTruthFilter. It is only used in the record of options passed to the function. The examples generated and used in the actual tests are records as well. There sourceOfTruth is always a matrix (either some MatricObj or a classical GAP matrix).